### DISCUSSION TITLE
fit_transform of combined umap

### DISCUSSION THREAD

**Discussion by rabbithooper** [lmcinnes/umap#623] [Q&A] [answered]
Hello,

Is there a way to get embeddings of combined umaps either union or intersection for downstream actions such as hdbscan?

**Comment by lmcinnes** [comment, marked as answer]
Sorry, I'm not quite sure what you are asking. If the question is about transform capabilities then the answer is no. If it is simply a matter of accessing the embedding then the answer is yes -- the ``embedding_`` attribute of the combined model contains that data.

**Reply by rabbithooper** [reply]
Thanks, was wondering about transform capabilities

**Reply by lmcinnes** [reply]
In principle such a thing is possible, but requires passing in new data from each dataset to be combined, and is a little tricky to get right. It hasn't been coded up yet. If you are particularly interested in such a thing we could discuss what would be involved in implementing it and you could potentially make a basic implementation for yourself.

**Reply by yusuftalhatamer** [reply]
> In principle such a thing is possible, but requires passing in new data from each dataset to be combined, and is a little tricky to get right. It hasn't been coded up yet. If you are particularly interested in such a thing we could discuss what would be involved in implementing it and you could potentially make a basic implementation for yourself.

@lmcinnes Can you elaborate a bit more on the implementation of this? I am also interested.

**Reply by lmcinnes** [reply]
The transform works by connecting the new samples to the existing simplicial structure (hereafter graph, since that will suffice fir these discussions). Since we are fixing the training data we only need concern ourselves with edges from the new data to the training data. Having constructed those with essentially the same calculations used in regular UMAP, we then optimize the low dimensional layout holding the positions of the training samples fixed and moving only the new samples.

To make this work for a combined dataset model one would need to construct edges from the new samples to the training samples in the combined graph, and calculate them appropriately. From there things could mostly proceed as before, so the real catch is in creating and weighting those edges appropriately.

For the combined model we create two graphs, one for each dataset, and then combine the two graphs via (fuzzy set) intersection, union, or complement-intersection across the fuzzy set of edge weights. There is then some finessing to re-establish some of our assumptions -- you can see the code for those details; things like ``reset_local_connectivity``. Once we have the single combined graph we proceed as normally.

Thus for a transform on the combined dataset model we need to take in samples that have data for *both* sides of the datasets that were combined so we can create edges from them to the training data for each side, and then combine the newly generated edges to get a result that looks like the regular transform with one graph with edges from the new samples. So far, so good. Since we are using nearest neighbor queries, however, things will be a little sparse and there is some danger of essentially having no edges added in the combined result. The best way around this is likely to calculate near neighbors in each dataset and then take the union of all those edges and calculate appropriate weights for those in each dataset independently (so that we always have two edges to combine) and then combine them. The finessing with ``reset_local_connectivty`` may also require some adjustments for this specific case.

If all that gets implemented, however, then a the rest of the transform phase should be able to proceed as normal.

That's at least a sketch of what all would need to be done. Hopefully it is a start if you are interested in looking into this further.
